### DISCUSSION TITLE
Trying to make Unpoly, AlpineJS and LeafletJS work

### DISCUSSION THREAD

**Discussion by marcus-at-localhost** [unpoly/unpoly#583] [Q&A] [answered]
Hi, I'm completely lost solving the following issue:

I've got an unpoly app mainly for the `up-follow` functionality. Navigating between pages/screens works well. I've setup a few compilers, that also works well.

Now I've got a more complex part of the app which is a reactive map (having reactive HTML Elements to filter POIs on a map)
For that I started writing an AlpineJS component, that handles setting up a leafletJS map and interact with it.

That all works, but I run in those well known issues of code being executed multiple times. The stuff that happens if you don't take care or better: understand the lifecycle of those JS libraries.

html:
```pug
head
	script(src="alpine.js", defer)
	script(src="unpoly.js")
	script(src="poiMapAlpine.js")
	script(src="app.js")
body
	div#map(x-ignore, x-data="switchboard", up-data="{}")
	div
		template(x-for="visibleMarker in visibleMarkers")
			div(x-text="visibleMarker.title")
```

app.js contains my compiler:

```js
up.link.config.followSelectors.push('a[href]');
up.link.config.instantSelectors.push('a[href]');
up.link.config.preloadSelectors.push('a[href]');

up.compiler('#map', function (element, data) {
    // init Alpine by removing x-ignore
    element.removeAttribute('x-ignore');
	// store the data from up-data in Allpine store
    Alpine.store('config', data);
	
	// initialize the rendering of the <template> elements
	Alpine.nextTick(() => {
		Alpine.initTree(element);
	});

    return () => {
        console.log('destroy compiler')
		// destroy AlpineComponent
    }
});
```

poiMapAlpine.js contains the following:

```js

document.addEventListener("alpine:init", () => {
    Alpine.data('switchboard', () => ({

        data: Alpine.store('config'), // we stored that in the compiler
        map: null,
        visibleMarkers: [],

        init() {
            console.log('poiMap.init()');
            this.map = L.map(o.container);

			// some leaflet marker voodoo

            this.getVisibleMarkers();
            this.map.on('moveend', () => {
                this.getVisibleMarkers();
            });
        },

		// render all visible marker in a list with Alpine.js template(x-for)
        getVisibleMarkers() {
            var bounds = this.map.getBounds();
            var visibleMarkers = [];
            this.pois.eachLayer(function(layer) {
                if (layer instanceof L.Marker) {
                    if (bounds.contains(layer.getLatLng())) {
                        visibleMarkers.push(layer.feature.properties);
                    }
                } else if (layer instanceof L.Polygon) {
                    if (bounds.intersects(layer.getBounds())) {
                        visibleMarkers.push(layer.feature.properties);
                    }
                }
            });

            this.visibleMarkers = JSON.parse(JSON.stringify(visibleMarkers));
        },

        iAmHere() {
            if(this.map) {
                this.map.setView(this.data.currentLatlng, 20);
            }
        },

        destroy() {
            console.log('poiMap.destroy()');
			// can't access this.map.remove() or something
        }
    }))
});
```

here is a screencast of what's going on in the console: https://www.dropbox.com/scl/fi/dyz3tk1on14f5wock6uup/22-19-56.mp4?rlkey=8uetlo58kl6db7ihdqpftjgqs&dl=0

You can see the map is initialized correctly, it renders a list of `visibleMarkers` with alpineJS `<template>` and when I navigate away the Unpoly `destroy` (destroy compiler) method is called, then Unpoly sets up the new page/fragment and then AlpineJS `destroy` method is called (poiMap.destroy()) and Alpine crashes while `doCleanup` - I assume because the `template` tags are not in the dom anymore, since Unpoly already replaced everything. 

In my nativity, I thought I could put something like `Alpine.destroy()` into Unpoly's destroy method, to tear that whole thing down and reinstate it again when Unpoly calls  the compiler on navigating `javascript:history.back()`. But there is no Alpine.destroy() that removes the Alpine object completely. Instead it stays available. And here I can't wrap my mind around it anymore.

I read [Alpines Lifecycle](https://alpinejs.dev/essentials/lifecycle) page up and down and their forum, hoping to find something about similar behavior from htmx or turbolinks, but I couldn't figure it out.

TL;DR - I don't expect anyone to understand or fix my AlpineJS problem, but I hope someone with more insight into Unpoly and the issues around DOM swapping and losing references to elements etc. can point me in the right direction. I assume a VueJS component would have the same issues? 

What concepts should I look into? Is there something I can setup with unpoly that prevents this behavior? (disable caching for the map? I know I was able to do that in turbolinks back then).

Thanks for any hints, and thanks for your time reading and trying to understand. I gave my best to describe it.

**Comment by triskweline** [comment]
Hi @marcus-at-localhost. I have no experience with Alpine, but here are a few thoughts:

- I'm not sure if `Alpine.data()` is meant to do DOM manipulation. Alpine docs use this function to initialize data and define some controller methods.
- I'm not sure if Alpine guarantees that the `Alpine.data()` callback only runs once.
- I don't understand the purpose of preventing Alpine.js initialization (through `x-ignore`), then having a compiler that does nothing, then initialize Alpine.js in the next tick.
- Have you considered leaving out Alpine, and just render your map using an `up.compiler()`? Pass the marker data in `[up-data]`, have the compiler render the `<template>`, then call Leaflet.

> disable caching for the map? I know I was able to do that in turbolinks back then

You can [disable caching](https://unpoly.com/caching#disabling-caching), but I don't see cached content being used in your video. Also when cached content is rendered, Unpoly recreates every element from the cached HTML. No DOM element is reused.

> I assume a VueJS component would have the same issues?

While there may always be issues when two libraries want to control the same DOM tree, we are using React components within Unpoly.js without problems.

> In my nativity, I thought I could put something like Alpine.destroy() into Unpoly's destroy method, to tear that whole thing down and reinstate it again when Unpoly calls the compiler on navigating javascript:history.back().

I believe Alpine registers a `MutationObserver` to automatically detect when elements are inserted into or removed from the DOM. You should not need to initialize or destroy Alpine state from Unpoly.

**Reply by marcus-at-localhost** [reply]
Thank you @triskweline for taking the time and share your thoughts. There are definitely some leads to explore further.
I thought about just using Unpoly and Leaflet because I know that works. But with Alpine or Vue Petite or similar, there is this promise of creating lean, reactive apps. But I guess at the end of the day it’s only used to build an accordion real fast.
Much respect for Unpoly! (Also loved to hear the Unpoly backstory at the Working Draft Podcast)

**Reply by marcus-at-localhost** [reply, marked as answer]
Ok, my approach was so convoluted. I'm going to write down where and why I lost the plot:

> I don't understand the purpose of preventing Alpine.js initialization (through x-ignore), then having a compiler that does nothing, then initialize Alpine.js in the next tick.

It came from this post :) https://github.com/unpoly/unpoly/discussions/314#discussioncomment-1635702 - I just threw everything at the wall to see what sticks ^^

> I believe Alpine registers a MutationObserver to automatically detect when elements are inserted into or removed from the DOM. **You should not need to initialize or destroy Alpine state from Unpoly.**

Correct! But since the Unpoly docs state:

> up.compiler() is a great way to integrate JavaScript libraries. Let's say your JavaScript plugin wants you to call lightboxify() on links that should open a lightbox.

I assumed AlpineJS is a 3rd Party library that should be called from a compiler. 😬

> Have you considered leaving out Alpine, and just render your map using an up.compiler()? Pass the marker data in [up-data], have the compiler render the <template>, then call Leaflet.



I removed the compiler, read the data-config attribute when initializing AlpineJS and ... it just works. 🤦

```pug
head
	script(src="unpoly.js")
	script(src="poiMapAlpine.js") // loading order matters!
	script(src="app.js")
	script(src="alpine.js", defer)
body
	div#map(x-ref="poiMap", x-data="switchboard", data-config="{ /* JSON Object*/ }")
	div
		template(x-for="visibleMarker in visibleMarkers")
			div(x-text="visibleMarker.title")
```

poiMapAlpine.js

```js
document.addEventListener("alpine:init", () => {
    Alpine.data('switchboard', () => ({

        data: null
        map: null,

        init() {
			this.data = JSON.parse(this.$refs.poiMap.dataset.config);
            this.map = L.map(o.container);
        }

       // the rest 
});
```

Thanks again @triskweline for taking the time and asking the right questions, this helped a lot.
